### PR TITLE
feat: log filters

### PR DESCRIPTION
**Pull request by jasonkuhrt**
closes graphql-nexus/nexus#264

BREAKING CHANGE:

- `level` is no longer a top-level setting. It is nested under `filter`. You can also specify `level` at the filter level. `level` setting is for setting the default filter level when none is given in a pattern.
- path names must not use valid JS identifier names. This is probably good to enforce anyways, but it functionally helps avoid collisions between path names  and pattern syntax.

#### TODO

- [x] ~docs~
- [x] tests
- [x] receive log filter from envar, test this
- [x] test that path names use safe identifier names that won't conflict with filter patterns
- [x] test filter processing (see todo)
- [x] syntax guide colored output
- [x] system tests

Example of the Syntax guide:

image

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,7 +14,7 @@
     "dev:test": "jest --watch",
     "build": "tsc",
     "test": "jest",
-    "clean": "rm -rf dist && git clean -fx",
```
Comment (jasonkuhrt) [inline package.json:17]: I screwed myself pretty bad with this. I lost a great test suite, hour+ of work. Very luckily I had the implementation open in VSCode and it was kept around (marked as deleted). I recovered it. Feel pretty stupid :D

File: src/level.ts [hunk 2]
```diff
@@ -1,38 +1,38 @@
-export type Level = 'fatal' | 'error' | 'warn' | 'info' | 'debug' | 'trace'
+export type Name = 'fatal' | 'error' | 'warn' | 'info' | 'debug' | 'trace'
 
-export type LevelNum = 60 | 50 | 40 | 30 | 20 | 10
+export type Num = 6 | 5 | 4 | 3 | 2 | 1
```
Comment (jasonkuhrt) [inline src/level.ts:3]: breaking change

File: src/root-logger.ts [hunk 3]
```diff
@@ -1,295 +1,38 @@
...
+  const settingsManager = ((newSettings) => {
+    settings(newSettings)
+    Object.assign(settingsManager, settings)
     return logger
-  }) as Settings
-
-  const state = { settings } as State
-  const loggerLink = Logger.create(state, [opts?.name ?? 'root'], {})
+  }) as SettingsManager
+  Object.assign(settingsManager, settings)
```
Comment (jasonkuhrt) [inline src/root-logger.ts:33]: Very unhappy with how this system/code looks. Open to refactoring further later.

File: src/settings.ts [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+     *
+     * @default
+     *
+     * Takes the first value found, searching in the following order:
+     *
+     *  1. LOG_LEVEL envar if set
+     *  2. 'info' if NODE_ENV envar set to 'production'
+     *  3. 'debug'
+     */
+    level?: Level.Name
```
Comment (jasonkuhrt) [inline src/settings.ts:98]: breaking change

**Comment by jasonkuhrt** [comment]
See test suite for what this PR really brings.

File: src/settings.ts [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+     * The current node process ID assigned by the operating system. Acquired
+     * via `process.pid`.
+     *
+     * @defualt `true` if NODE_ENV="production"
+     */
+    pid?: boolean
+    /**
+     * The host name of the machine this process is running on. Acquired via `OS.hostname()`.
+     *
+     * @defualt `true` if NODE_ENV="production"
```
Comment (Weakky) [inline src/settings.ts:173]: typo in `defualt` -> `default`

File: src/filter.ts [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+    ${subtitle(`Lists`)}
+    app,nexus   ${subtle(`app and nexus paths at default level`)}
+
+    ${subtitle(`Levels`)}
+    *@info      ${subtle(`all paths at info level`)}
+    *@error-    ${subtle(`all paths at error level or lower`)}
+    *@debug+    ${subtle(`all paths at debug level or higher`)}
+    *@3         ${subtle(`all paths at info level`)}
+    *@4-        ${subtle(`all paths at error level or lower`)}
+    *@2+        ${subtle(`all paths at debug level or higher`)}
```
Comment (Weakky) [inline src/filter.ts:280]: So, `LOG_FILTER` can be used to set `LOG_LEVEL`, right?

Comment (jasonkuhrt) [reply src/filter.ts:280]: Yeah, there's this somewhat awkward balance between defaults and the filter. The filter is the source of truth but when things are left unspecified where should the the value be taken from? In other words `LOG_LEVEL` could be renamed now to `LOG_FILTER_DEFUALT_LEVEL`. But I'm not sure about all this yet.

**Pull request merged**